src/execution/operator/join/physical_iejoin.cpp

```diff
@@ -0,0 +1,1415 @@
+		auto right = cond.right->Copy();
+		auto sense = OrderType::INVALID;
+
+		// 2. if (op1 ∈ {>, ≥}) sort L1 in descending order
+		// 3. else if (op1 ∈ {<, ≤}) sort L1 in ascending order
+		// 4. if (op2 ∈ {>, ≥}) sort L2 in ascending order
+		// 5. else if (op2 ∈ {<, ≤}) sort L2 in descending order
+		switch (cond.comparison) {
+		case ExpressionType::COMPARE_GREATERTHAN:
+		case ExpressionType::COMPARE_GREATERTHANOREQUALTO:
+			sense = i ? OrderType::ASCENDING : OrderType::DESCENDING;
+			break;
+		default:
+			sense = i ? OrderType::DESCENDING : OrderType::ASCENDING;
+			break;
+		}
+		lhs_orders[i].emplace_back(BoundOrderByNode(sense, OrderByNullType::NULLS_LAST, move(left)));
+		rhs_orders[i].emplace_back(BoundOrderByNode(sense, OrderByNullType::NULLS_LAST, move(right)));
+	}
+
+	for (idx_t i = 2; i < conditions.size(); ++i) {
+		auto &cond = conditions[i];
+		D_ASSERT(cond.left->return_type == cond.right->return_type);
+		join_key_types.push_back(cond.left->return_type);
+	}
+
+	children.push_back(move(left));
+	children.push_back(move(right));
+}
+
+//===--------------------------------------------------------------------===//
+// Sink
+//===--------------------------------------------------------------------===//
+class IEJoinLocalState : public LocalSinkState {
+public:
+	explicit IEJoinLocalState(const vector<JoinCondition> &conditions, const idx_t child) : has_null(0), count(0) {
+		// Initialize order clause expression executor and key DataChunk
+		vector<LogicalType> types;
+		for (const auto &cond : conditions) {
+			comparisons.emplace_back(cond.comparison);
+
+			const auto &expr = child ? cond.right : cond.left;
+			executor.AddExpression(*expr);
+
+			types.push_back(expr->return_type);
+		}
+		keys.Initialize(types);
+	}
+
+	//! The local sort state
+	LocalSortState local_sort_state;
+	//! Local copy of the sorting expression executor
+	ExpressionExecutor executor;
+	//! Holds a vector of incoming sorting columns
+	DataChunk keys;
+	//! The comparison list (for null merging)
+	vector<ExpressionType> comparisons;
+	//! The number of NULL values
+	idx_t has_null;
+	//! The total number of rows
+	idx_t count;
+
+	idx_t MergeKeyNulls();
+
+	void Sink(DataChunk &input, GlobalSortState &global_sort_state) {
+		// Initialize local state (if necessary)
+		if (!local_sort_state.initialized) {
+			local_sort_state.Initialize(global_sort_state, global_sort_state.buffer_manager);
+		}
+
+		// Obtain sorting columns
+		keys.Reset();
+		executor.Execute(input, keys);
+
+		// Count the NULLs so we can exclude them later
+		has_null += MergeKeyNulls();
+		count += keys.size();
+
+		// Sink the data into the local sort state
+		if (keys.ColumnCount() > 1) {
+			//	Only sort the primary key
+			DataChunk join_head;
+			join_head.data.emplace_back(Vector(keys.data[0]));
+			join_head.SetCardinality(keys.size());
+
+			local_sort_state.SinkChunk(join_head, input);
+		} else {
+			local_sort_state.SinkChunk(keys, input);
```

Comment:
This appears to be untested

Reply:
In the back of my mind is a later unification of the utilities in IEJoin, PWMJ and maybe a planar sweep algorithm, so the merged code would use this branch, but I will remove it for now.